src/Native/System.Net.Security.Native/pal_ntlmapi.cpp

```diff
@@ -0,0 +1,267 @@
+    }
+    *outLength = static_cast<int32_t>((*bufferHandle)->length);
+    return status;
+}
+
+extern "C" void NetSecurity_HeimNtlmFreeBuf(ntlm_buf* data)
+{
+    assert(data != nullptr);
+    heim_ntlm_free_buf(data);
+    delete data;
+}
+
+extern "C" void NetSecurity_ExtractNtlmBuffer(const ntlm_buf* bufferHandle, uint8_t* destination, uint32_t capacity, uint32_t offset)
+{
+    if (bufferHandle == nullptr)
+    {
+        return;
+    }
+    assert(destination != nullptr);
+    assert(bufferHandle->length <= (capacity - offset));
+    memcpy(destination + UnsignedCast(offset), bufferHandle->data, bufferHandle->length);
+}
+
+extern "C" int32_t NetSecurity_HeimNtlmEncodeType1(uint32_t flags, ntlm_buf** outBufferHandle, int32_t* outLength)
+{
+    assert(outBufferHandle != nullptr);
+    ntlm_type1 type1;
+    memset(&type1, 0, sizeof(ntlm_type1));
+    type1.flags = flags;
+    *outBufferHandle = new ntlm_buf();
+    return NetSecurity_SetBufferLength(heim_ntlm_encode_type1(&type1, *outBufferHandle), outBufferHandle, outLength);
+}
+
+extern "C" int32_t NetSecurity_HeimNtlmDecodeType2(uint8_t* data, int32_t offset, int32_t count, ntlm_type2** type2)
+{
+    assert(data != nullptr);
+    assert(offset >= 0);
+    assert(type2 != nullptr);
+    ntlm_buf buffer{.length = UnsignedCast(count), .data = data + offset};
+    *type2 = new ntlm_type2();
+    int32_t stat = heim_ntlm_decode_type2(&buffer, *type2);
+    if (stat != 0)
+    {
+        delete *type2;
+        *type2 = nullptr;
+    }
+    return stat;
+}
+
+extern "C" void NetSecurity_HeimNtlmFreeType2(ntlm_type2* type2)
+{
+    assert(type2 != nullptr);
+    heim_ntlm_free_type2(type2);
+    delete type2;
+}
+
+extern "C" int32_t NetSecurity_HeimNtlmNtKey(const char* password, ntlm_buf** outBufferHandle, int32_t* outLength)
+{
+    assert(outBufferHandle != nullptr);
+    *outBufferHandle = new ntlm_buf();
+    return NetSecurity_SetBufferLength(heim_ntlm_nt_key(password, *outBufferHandle), outBufferHandle, outLength);
+}
+
+extern "C" int32_t NetSecurity_HeimNtlmCalculateResponse(int32_t isLM,
+                                                         const ntlm_buf* key,
+                                                         ntlm_type2* type2,
+                                                         char* username,
+                                                         char* target,
+                                                         uint8_t* baseSessionKey,
+                                                         int32_t baseSessionKeyLen,
+                                                         ntlm_buf** outBufferHandle,
+                                                         int32_t* outLength)
+{
+    // reference doc: http://msdn.microsoft.com/en-us/library/cc236700.aspx
+    assert(baseSessionKeyLen == MD5_DIGEST_LENGTH);
+    assert(isLM == 0 || isLM == 1);
+    assert(type2 != nullptr);
+    assert(key != nullptr);
+    assert(outBufferHandle != nullptr);
```

Comment:
Assert outLength is not nullptr.

(Please go through this whole file, and for every pointer parameter, assert/check that it isn't nullptr; or give a comment explaining why/when it can be nullptr)